Mathlib/Topology/Germ.lean

```diff
@@ -0,0 +1,190 @@
+the corresponding germ of functions `X → Z` at `x ∈ X` resp. `Y → Z` at `y ∈ Y`..
+* `eq_of_germ_isConstant`: if each germ of `f : X → Y` is constant and `X` is pre-connected,
+`f` is constant.
+-/
+
+variable {F G : Type*} [NormedAddCommGroup F] [NormedSpace ℝ F]
+  [NormedAddCommGroup G] [NormedSpace ℝ G]
+
+open scoped Topology
+
+open Filter Set
+
+variable {X Y Z : Type*} [TopologicalSpace X] {f g : X → Y} {A : Set X} {x : X}
+
+namespace Filter.Germ
+
+/-- The value associated to a germ at a point. This is the common value
+shared by all representatives at the given point. -/
+def value {X α : Type*} [TopologicalSpace X] {x : X} (φ : Germ (𝓝 x) α) : α :=
+  Quotient.liftOn' φ (fun f ↦ f x) fun f g h ↦ by dsimp only; rw [Eventually.self_of_nhds h]
+
+theorem value_smul {α β : Type*} [SMul α β] (φ : Germ (𝓝 x) α)
+    (ψ : Germ (𝓝 x) β) : (φ • ψ).value = φ.value • ψ.value :=
+  Germ.inductionOn φ fun _ ↦ Germ.inductionOn ψ fun _ ↦ rfl
+
+/-- The map `Germ (𝓝 x) E → E` as a monoid homeomorphism -/
+@[to_additive "The map `Germ (𝓝 x) E → E` as an additive monoid homeomorphism"]
+def valueMulHom {X E : Type*} [Monoid E] [TopologicalSpace X] {x : X} : Germ (𝓝 x) E →* E where
+  toFun := Filter.Germ.value
+  map_one' := rfl
+  map_mul' φ ψ := Germ.inductionOn φ fun _ ↦ Germ.inductionOn ψ fun _ ↦ rfl
+
+/-- The map `Germ (𝓝 x) E → E` into a `𝕜`-module `E` as a `𝕜`-linear map -/
+def valueₗ {X 𝕜 E : Type*} [Semiring 𝕜] [AddCommMonoid E] [Module 𝕜 E] [TopologicalSpace X]
+    {x : X} : Germ (𝓝 x) E →ₗ[𝕜] E :=
+  { Filter.Germ.valueAddHom with map_smul' := fun _ φ ↦ Germ.inductionOn φ fun _ ↦ rfl }
+
+/-- The map `Germ (𝓝 x) E → E` as a ring homeomorphism -/
+def valueRingHom {X E : Type*} [Semiring E] [TopologicalSpace X] {x : X} : Germ (𝓝 x) E →+* E :=
+  { Filter.Germ.valueMulHom, Filter.Germ.valueAddHom with }
+
+/-- The map `Germ (𝓝 x) E → E` as a monotone ring homeomorphism -/
+def valueOrderRingHom {X E : Type*} [OrderedSemiring E] [TopologicalSpace X] {x : X} :
+    Germ (𝓝 x) E →+*o E where
+  __ := Filter.Germ.valueRingHom
+  monotone' := fun φ ψ ↦
+  Germ.inductionOn φ fun _ ↦ Germ.inductionOn ψ fun _ h ↦ h.self_of_nhds
+
+end Filter.Germ
+
+/-- The inclusion `S → R` of a subring, as an ordered ring homomorphism. -/
+-- xxx: OrderedRing has no morphisms, OrderedRingHom no subtypes -> which file is a good place?
+def _root_.Subring.orderedSubtype {R} [OrderedRing R] (s : Subring R) : s →+*o R :=
+  { s.subtype with monotone' := fun _ _ h ↦ h }
+
+section RestrictGermPredicate
+/-- Given a predicate on germs `P : Π x : X, germ (𝓝 x) Y → Prop` and `A : set X`,
+build a new predicate on germs `restrict_germ_predicate P A` such that
+`(∀ x, restrict_germ_predicate P A x f) ↔ ∀ᶠ x near A, P x f`, see
+`forall_restrict_germ_predicate_iff` for this equivalence. -/
+def RestrictGermPredicate (P : ∀ x : X, Germ (𝓝 x) Y → Prop)
+    (A : Set X) : ∀ x : X, Germ (𝓝 x) Y → Prop := fun x φ ↦
+  Quotient.liftOn' φ (fun f ↦ x ∈ A → ∀ᶠ y in 𝓝 x, P y f)
```

Comment:
This should use [Filter.Germ.liftOn](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Filter/Germ.html#Filter.Germ.liftOn) instead of `Quotient.liftOn'`